Weird categories appearance fix + other

AndySky21 opened the pull request:
Fixed the bug discussed in "Issues". Also replaced the document.write with direct markup, as the extension is useful also in no-js scenarios. Polyglot markup for the table.
Removed CSS expression, as it's just for old versions of IE. Also, CSS rules are now all in the stylesheet.

adamfranco commented on CategorySuggest.body.php:
```diff
@@ -95,17 +90,13 @@ function fnCategorySuggestSaveHook( $m_isUpload, $m_pageObj ) {
 	 	 	if(strlen($m_cat)>0){
 				$m_cat = Title::capitalize($m_cat, NS_CATEGORY);
-				$m_text .= "\n[[$m_catString:" . trim($m_cat) . "]]";
+				$m_text .= "\n[[". $m_catString .":" . mysql_escape_string(trim($m_cat)) . "]]";
```
@AndySky21 , pull #18 looks like some nice cleanup. I haven't had a chance to test yet, but a cursory review only turned up one problem: the addition of `mysql_escape_string()` I'm not sure if this was coming in from a merge-resolution with master and can just be removed, but if not we'll need to find a different way to accomplish the goal as not everyone uses MySQL (we had bug reports mentioning running MediaWiki on SQLITE and Postgres as well).

AndySky21 replied:
I don't really understand where that function comes from. Additionally, mysql functions are deprecated in PHP 5 too and here it has nothing to do. Removed.

AndySky21 commented:
found a bug in page save. Please wait until fixed.

AndySky21 commented:
The corrections on CategorySuggest.body.php now also include a faster version of the restricted tag recognition system, as well as support for <onlyinclude> tag. It also resolves the bug that tags nested inside <nowiki> (with the obvious exception of </nowiki>) do not stack in/out, as they aren't parsed at all.

AndySky21 commented:
Edit: improved fnCategorySuggestGetPageCategories by recycling $m_pageText as array

AndySky21 commented:
Ok this was last. There's still a bug in that extension CSS doesn't show up in upload page, but it's a special page so I don't think much can be done.

adamfranco commented:
Thanks, I'll review and test soon.

adamfranco commented:
Hi @AndySky21. Thanks for continuing to work on this. I noticed that commit ba72e3fc introduces a new global, `$foundCategories`. Reading through the code, it's unclear to me what the goal was in changing from:

    function fnCategorySuggestGetPageCategories( $m_pageObj ) {
        ...
        $outputText .= fnCategorySuggestStripCats($block, $foundCategories);
        ...
    }
    function fnCategorySuggestStripCats($texttostrip, &$catsintext){
        ...
        array_push($catsintext,$newString);
        ...
    }

to:

    function fnCategorySuggestGetPageCategories( $m_pageObj ) {
        ...
        $outputText .= fnCategorySuggestStripCats($block,);
        ...
    }
    function fnCategorySuggestStripCats($texttostrip){
        global $foundCategories
        ...
        $foundCategories[] = ucfirst($matches[2]);
        ...
    }

As a rule, I'd prefer the code to not introduce any globals for things other than configuration or built-in MediaWiki state as this makes it less clear what the inputs of the function are and the order of operations. Please switch back to passing in the category-list-array by reference. Thanks! :smile:

AndySky21 commented:
Structure fixed! It took some time to me in order to figure out how to pass a variable by reference in an internal anonymous callback function.
You can see the result now. Also, I removed some unused features. Look at the commit summary to see what I removed, and tell me if you were planning to use that stuff in future improvements.

AndySky21 commented:
One more note: as of PHP 5.3, call-time pass-by-reference raises a warning. As of PHP5.4, it is considered fatal error. Variables can be passed by reference only in the function definition. Commit 13b4d8e takes it into account.

adamfranco commented:
Ah, passing by reference into an anonymous callback isn't something I've needed to do before. Thanks for figuring it out!

I just did some testing of this branch and ran into two additional errors that will need to be addressed:

**Search list overlay problems**
First, the CSS changes have now broken the floating-overlay behavior of search-results list-div. Here's a video of how it used to work: [Category-suggest-orig.mp4](http://middmedia.middlebury.edu/middmedia/view/dir/afranco/file/Category-suggest-orig.mp4), and here's a video of how it is working now: [Category-suggest-pull18.mp4](http://middmedia.middlebury.edu/middmedia/view/dir/afranco/file/Category-suggest-pull18.mp4). I wasn't able to figure out what the goal of the CSS changes included in this pull request were, but if the old behavior can't be replicated with the changes, they should be removed.

**unset() causing Undefined index notices**
Using the `for` loop in `fnCategorySuggestGetPageCategories($m_pageObj)` and then unsetting elements in the array is causing notices that break edit-page rendering. I tried loading the edit page with the following page-content:

	:<div class="notice metadata plainlinks" id="stub" style="clear:both;">''This article is a [[:Category:Stubs|stub]]. You can help
	{{SITENAME}} by  [
	{{fullurl:{{FULLPAGENAME}}|action=edit}} expanding it].
	{{{1|}}}''</div>
	<includeonly>[[Category:Stubs]]</includeonly>
	<noinclude>
	----
	Usage
	* Add the text <nowiki>{{stub}}</nowiki> to the top of your page.
	* You can also include an optional message: <nowiki>{{stub|Hello World}}</nowiki>
	About
	*This template will categorize articles that include it into [[:Category:Stubs]].
	[[Category:Article management templates|{{PAGENAME}}]]
	</noinclude>

and got the following notices which caused other warnings about page-headers already being sent:

    Notice: Undefined offset: 2 in [...]/CategorySuggest/CategorySuggest.body.php on line 136
    Notice: Undefined offset: 3 in [...]/CategorySuggest/CategorySuggest.body.php on line 136
    Notice: Undefined offset: 6 in [...]/CategorySuggest/CategorySuggest.body.php on line 136
    ...

The way this should be fixed is to change the parenthetical statements in the regular expression from `( ... )` to `(?: ... )`. The `?:` tells the regular expression that this parenthetical should be used for grouping `OR` options, but not captured and returned in the matched output. The new line 135 should be:

    $m_pageText = preg_split('#(</?(?:'. implode('|',$reservedTags) . ')>|(?:\{\{|\}\}))#u', $m_pageText , null, PREG_SPLIT_DELIM_CAPTURE);

With this change in place you can get rid of all of the `unset()` statements and convert the loop back to a simple foreach loop.

adamfranco commented:
P.S. we're getting close on all of these changes being ready to go, but in the future creating a new branch with a separate pull-request for each independent improvement, such as one for the css changes, one for the javascript reworking, one for tag-exclusion reworking, etc might make it easier to review and improve them. :smile:

AndySky21 commented:
@adamfranco : for instance, thank you for the suggestions about how to work on projects on GitHub. Really. This is the second collaborative I contribute to, so I'm still a little unconfident about how to manage work. And I'm not used to teamwork, so everything is worse. Moreover, things to fix appeared almost randomly so I fixed them in my local Wiki project and then submitted it to my fork.
Thus said, I didn't know about the ?: statement in regex. Good to know that not every parenthesis must end up in a reference. I'll fix it now, as it is for CSS. Of course the problem is not in the rules being moved in the stylesheet, but rather in some default rule not fixed. I'll write some lines here when it's done.

AndySky21 commented:
The error on CSS rules was something old I didn't remove on the GitHub file. Now it works.
Fixed errors in body.php: now groups don't appear in references.
Any ideas about how to integrate CSS on special:upload?

adamfranco commented:
`fnCategorySuggestGetPageCategories` looks good. Thanks for updating.

I just tested again with your latest changes and the search-results list is still expanding the whole `<div id="categoryselectmaster">` element rather than leaving that div with its original size and floating the results over the lower part of the page. Here's a video of the latest state:
[Category-suggest-pull18-2b182f78.mp4](http://middmedia.middlebury.edu/middmedia/view/dir/afranco/file/Category-suggest-pull18-2b182f78.mp4). It *should* have the same display behavior as the old version: [Category-suggest-orig.mp4](http://middmedia.middlebury.edu/middmedia/view/dir/afranco/file/Category-suggest-orig.mp4). Notice in the original how the Category Suggest form and the Save buttons below all stay in exactly the same position and same height and the results element just floats over top.

I don't have any ideas off hand how to include the CSS on the Special:Upload page. Given that the Special:Upload CSS issue was preexisting is separate from the rest of the work on these improvements, how about creating a separate issue for it with a separate branch/pull-request if you figure it out. :smile:

AndySky21 commented:
It was a matter of CSS priority. It should be ok now.

adamfranco commented:
Thanks for all your work on this improvement! It is now merged. :smile: 

If you have future improvements or fixes, here is a quick overview:

1. [Optional] Create a new issue for the fix. This isn't always required (Pull requests alone are OK), but can be handy if you won't be able to get around to making the fix right away. Maybe someone else can jump on it.
2. Create a new branch for the new fix/improvement. Usually it's best to create the branch at the most recent commit in the main project's master branch, but this isn't strictly necessary.
3. Make the improvement and commit your changes to your branch. If there is an issue related to this work, you can mention the issue number in the commit message to attach a record of the commit to the issue. For example: `Fix CSS overlay issue in #18`.
4. Push your new branch to your Github fork and create a pull request.
5. Once the project maintainer merges the pull request you can optionally delete your feature-branch.

The GitHub help section has some great info on this workflow:
* https://guides.github.com/introduction/flow/index.html
* https://help.github.com/articles/github-flow-in-the-browser/
* http://scottchacon.com/2011/08/31/github-flow.html

By keeping each feature separate we can merge in one set of changes right away and continue to work through more sticky issues or regressions on others.

Best,
Adam

adamfranco commented:
One final note. Follow this guide for commit messages: http://chris.beams.io/posts/git-commit/ 

* Write a descriptive subject line
* Separate subject from body with a blank line
* Limit the subject line to 50 characters
* Capitalize the subject line
* Wrap the body at 72 characters
* Use the body to explain what and why vs. how

There isn't a need to mention the filename that was changed as this is obvious from the diff. :wink:

AndySky21 commented:
Thank you very much for these suggestions! I'm now going to submit a different pull request for my solution to the Special:Upload issue, from the branch I created. It isn't that neat, so feel free to discard it.
Best,
Andrea
